onert-micro/luci-interpreter/src/core/RuntimeGraph.cpp

```diff
@@ -89,83 +143,113 @@ void RuntimeGraph::deallocate(size_t kernel_index) const
   }
 }
 
-RuntimeGraph::RuntimeGraph(RuntimeModule *owning_module, IMemoryManager *memory_manager)
-  : _owning_module(owning_module), _memory_manager(memory_manager)
+void RuntimeGraph::configure()
 {
-}
+  if (not _is_valid)
+    buildAllocDeallocPlan();
 
-RuntimeGraph::~RuntimeGraph()
-{
-  for (auto &tensor : _tensors)
+  for (auto &kernel : _kernels)
   {
-    if (tensor->is_data_allocated())
-      _memory_manager->release_memory(*tensor);
+    kernel->configure();
   }
-}
 
-Tensor *RuntimeGraph::addTensor(std::unique_ptr<Tensor> &&tensor)
-{
-  assert(tensor != nullptr);
-  _tensors.push_back(std::move(tensor));
-  return _tensors.back().get();
+  _is_valid = true;
 }
 
-AffineQuantization *
-RuntimeGraph::addAffineQuantization(std::unique_ptr<AffineQuantization> &&quantization)
+void RuntimeGraph::execute()
 {
-  assert(quantization != nullptr);
-  _affine_quantizations.push_back(std::move(quantization));
-  return _affine_quantizations.back().get();
-}
+  if (not _is_valid)
+    configure();
 
-void RuntimeGraph::addInputTensor(Tensor *input_tensor) { _input_tensors.push_back(input_tensor); }
+  for (size_t index = 0; index < _kernels.size(); ++index)
+  {
+    const auto &kernel = _kernels[index];
 
-void RuntimeGraph::addOutputTensor(Tensor *output_tensor)
-{
-  _output_tensors.push_back(output_tensor);
+    // TODO: add kernel->configure for methods with dynamic shapes
+
+    // Preallocate outputs in advance instead of relying on automatic allocation
+    allocate(index);
+
+    kernel->execute();
+
+    deallocate(index);
+  }
 }
 
-void RuntimeGraph::configureAllocations(Tensor *tensor)
+// StaticRuntimeGraph
+StaticRuntimeGraph::StaticRuntimeGraph(IMemoryManager *memory_manager)
+  : IBaseRuntimeGraph(memory_manager)
 {
-  _memory_manager->allocate_memory(*tensor);
 }
 
-void RuntimeGraph::addKernel(std::unique_ptr<Kernel> &&kernel)
+StaticRuntimeGraph::~StaticRuntimeGraph()
 {
-  assert(kernel != nullptr);
-  _kernels.push_back(std::move(kernel));
-  _is_valid = false;
+  // Release intermediate computing buffer.
+  const auto static_memory_manager = dynamic_cast<StaticMemoryManager *>(_memory_manager);
```

Comment:
Do we use dynamic casts in other parts of interpreter?
I think it is a good idea to omit such RTTI if possible